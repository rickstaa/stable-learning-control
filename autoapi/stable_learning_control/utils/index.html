<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>stable_learning_control.utils &mdash; stable-learning-control 6.0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../../../_static/graphviz.css?v=eafc0fe6" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/modify.css?v=519ed47b" />

  
    <link rel="shortcut icon" href="../../../_static/favicon.ico"/>
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js?v=9497d378"></script>
        <script src="../../../_static/doctools.js?v=888ff710"></script>
        <script src="../../../_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="stable_learning_control.utils.log_utils" href="log_utils/index.html" />
    <link rel="prev" title="stable_learning_control.disturbers.observation_random_noise_disturber" href="../disturbers/observation_random_noise_disturber/index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html">
            
              <img src="../../../_static/logo.svg" class="logo" alt="Logo"/>
          </a>
              <div class="version">
                6.0.1
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Usage</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../usage/installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../usage/docker.html">Use with Docker</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../usage/algorithms.html">Available Agents</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../usage/running.html">Running Experiments</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../usage/hyperparameter_tuning.html">Hyperparameter Tuning</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../usage/saving_and_loading.html">Experiment Outputs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../usage/plotting.html">Plotting Results</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../usage/eval_robustness.html">Evaluating Robustness</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../usage/benchmarks.html">Benchmarks</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Utilities</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../utils/loggers.html">Loggers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../utils/mpi.html">MPI Tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../utils/run_utils.html">Run Utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../utils/plotter.html">Plotter</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../utils/testers.html">Policy testers</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Development</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../dev/contributing.html">Contribute to stable-learning-control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../dev/doc_dev.html">Build the documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../dev/license.html">License</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Documentation</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../../index.html">API Reference</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../index.html">stable_learning_control</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="../index.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="../algos/index.html">stable_learning_control.algos</a></li>
<li class="toctree-l4"><a class="reference internal" href="../common/index.html">stable_learning_control.common</a></li>
<li class="toctree-l4"><a class="reference internal" href="../disturbers/index.html">stable_learning_control.disturbers</a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#">stable_learning_control.utils</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../index.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="../index.html#attributes">Attributes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../index.html#functions">Functions</a></li>
<li class="toctree-l3"><a class="reference internal" href="../index.html#package-contents">Package Contents</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Etc.</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../etc/acknowledgements.html">Acknowledgements</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">stable-learning-control</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">API Reference</a></li>
          <li class="breadcrumb-item"><a href="../index.html">stable_learning_control</a></li>
      <li class="breadcrumb-item active">stable_learning_control.utils</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/rickstaa/stable-learning-control/blob/main/docs/source/autoapi/stable_learning_control/utils/index.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-stable_learning_control.utils">
<span id="stable-learning-control-utils"></span><h1>stable_learning_control.utils<a class="headerlink" href="#module-stable_learning_control.utils" title="Permalink to this heading"></a></h1>
<p>Contains all the SLC utilities.</p>
<section id="subpackages">
<h2>Subpackages<a class="headerlink" href="#subpackages" title="Permalink to this heading"></a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="log_utils/index.html">stable_learning_control.utils.log_utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="mpi_utils/index.html">stable_learning_control.utils.mpi_utils</a></li>
</ul>
</div>
</section>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this heading"></a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="eval_robustness/index.html">stable_learning_control.utils.eval_robustness</a></li>
<li class="toctree-l1"><a class="reference internal" href="eval_utils/index.html">stable_learning_control.utils.eval_utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="gym_utils/index.html">stable_learning_control.utils.gym_utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="import_utils/index.html">stable_learning_control.utils.import_utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="plot/index.html">stable_learning_control.utils.plot</a></li>
<li class="toctree-l1"><a class="reference internal" href="run_entrypoint/index.html">stable_learning_control.utils.run_entrypoint</a></li>
<li class="toctree-l1"><a class="reference internal" href="run_utils/index.html">stable_learning_control.utils.run_utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="safer_eval_util/index.html">stable_learning_control.utils.safer_eval_util</a></li>
<li class="toctree-l1"><a class="reference internal" href="serialization_utils/index.html">stable_learning_control.utils.serialization_utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="test_policy/index.html">stable_learning_control.utils.test_policy</a></li>
</ul>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../disturbers/observation_random_noise_disturber/index.html" class="btn btn-neutral float-left" title="stable_learning_control.disturbers.observation_random_noise_disturber" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="log_utils/index.html" class="btn btn-neutral float-right" title="stable_learning_control.utils.log_utils" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Rick Staa.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>