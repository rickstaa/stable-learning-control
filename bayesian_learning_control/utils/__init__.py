"""Contains several utilities that can be used across all the BLC modules.
"""
