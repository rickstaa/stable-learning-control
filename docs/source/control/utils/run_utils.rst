=========
Run Utils
=========

.. contents:: Table of Contents

ExperimentGrid
==============

Spinning Up ships with a tool called ExperimentGrid for making hyperparameter ablations easier. This is based on (but simpler than) `the rllab tool`_ called VariantGenerator.

.. _`the rllab tool`: https://github.com/rll/rllab/blob/master/rllab/misc/instrument.py#L173

.. autoclass:: machine_learning_control.control.utils.run_utils.ExperimentGrid
    :members:


Calling Experiments
===================

.. autofunction:: machine_learning_control.control.utils.run_utils.call_experiment

.. autofunction:: machine_learning_control.control.utils.log.setup_logger_kwargs
