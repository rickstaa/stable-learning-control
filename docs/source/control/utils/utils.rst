=========
Utilities
=========

The :mlc:`Machine Learning Control <>` contains several Utilities for


.. toctree::
   :maxdepth: 2
   :caption: Control

   ../control
   ../algorithms/sac
   ../running
   ../saving_and_loading
   ../plotting
   ../bench/bench