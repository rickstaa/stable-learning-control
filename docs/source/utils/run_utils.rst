=========
Run Utils
=========

.. _runner:

Command Line Interface
======================

The SLC package contains a command line interface (CLI) that can invoke several run utilities directly from the command line. You
can use the following command to check which utilities are available quickly:

.. code-block:: bash

    python -m stable_learning_control.run --help

The SLC package currently contains the following utilities:

+-----------------------+------------------------------------------------------------------------------------------------------------------------------+
| Utility               | description                                                                                                                  |
+=======================+==============================================================================================================================+
| :ref:`test_policy`    | Utility used to evaluate the performance of the trained policy in a given environment.                                       |
+-----------------------+------------------------------------------------------------------------------------------------------------------------------+
| :ref:`plot`           | Utility used to plot diagnostics from experiments.                                                                           |
+-----------------------+------------------------------------------------------------------------------------------------------------------------------+
| :ref:`eval_robustness`| Utility used to evaluate the robustness of a trained policy against external disturbances and changes in initial conditions. |
+-----------------------+------------------------------------------------------------------------------------------------------------------------------+

See the corresponding module documentation for more information on a given utility.

.. autofunction:: stable_learning_control.run.run

.. _exp_grid_utility:

ExperimentGrid utility
======================

SLC ships with a tool called ExperimentGrid for making hyperparameter ablations easier. This is based on (but simpler than) `the rllab tool`_ called VariantGenerator.

.. _`the rllab tool`: https://github.com/rll/rllab/tree/master/rllab/misc/instrument.py#L173

.. autoclass:: stable_learning_control.utils.run_utils.ExperimentGrid
    :members:

.. _exp_call_utility:

Calling Experiments utility
===========================

.. autofunction:: stable_learning_control.utils.run_utils.call_experiment
