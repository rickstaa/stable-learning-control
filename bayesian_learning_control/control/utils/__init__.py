"""A set of common utilities used within the algorithm code.
"""

from bayesian_learning_control.control.utils.safer_eval import safer_eval
