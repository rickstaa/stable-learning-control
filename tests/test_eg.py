"""This script is used to test whether the ExperimentGrid class is working as expected.
"""

from bayesian_learning_control.control.utils.run_utils import ExperimentGrid


# Check if the ExperimentGrid class is still working as expected.
def test_eg(snapshot):
    eg = ExperimentGrid()
    eg.add("test:a", [1, 2, 3], "ta", True)
    eg.add("test:b", [1, 2, 3])
    eg.add("some", [4, 5])
    eg.add("why", [True, False])
    eg.add("huh", 5)
    eg.add("no", 6, in_name=True)

    # Ensure that the variant is equal to the snapshot.
    assert eg.variants() == snapshot
