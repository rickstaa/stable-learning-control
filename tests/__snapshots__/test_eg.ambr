# serializer version: 1
# name: test_eg
  list([
    dict({
      'huh': 5,
      'no': 6,
      'some': 4,
      'test': dict({
        'a': 1,
        'b': 1,
      }),
      'why': True,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 4,
      'test': dict({
        'a': 1,
        'b': 1,
      }),
      'why': False,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 5,
      'test': dict({
        'a': 1,
        'b': 1,
      }),
      'why': True,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 5,
      'test': dict({
        'a': 1,
        'b': 1,
      }),
      'why': False,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 4,
      'test': dict({
        'a': 1,
        'b': 2,
      }),
      'why': True,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 4,
      'test': dict({
        'a': 1,
        'b': 2,
      }),
      'why': False,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 5,
      'test': dict({
        'a': 1,
        'b': 2,
      }),
      'why': True,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 5,
      'test': dict({
        'a': 1,
        'b': 2,
      }),
      'why': False,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 4,
      'test': dict({
        'a': 1,
        'b': 3,
      }),
      'why': True,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 4,
      'test': dict({
        'a': 1,
        'b': 3,
      }),
      'why': False,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 5,
      'test': dict({
        'a': 1,
        'b': 3,
      }),
      'why': True,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 5,
      'test': dict({
        'a': 1,
        'b': 3,
      }),
      'why': False,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 4,
      'test': dict({
        'a': 2,
        'b': 1,
      }),
      'why': True,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 4,
      'test': dict({
        'a': 2,
        'b': 1,
      }),
      'why': False,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 5,
      'test': dict({
        'a': 2,
        'b': 1,
      }),
      'why': True,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 5,
      'test': dict({
        'a': 2,
        'b': 1,
      }),
      'why': False,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 4,
      'test': dict({
        'a': 2,
        'b': 2,
      }),
      'why': True,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 4,
      'test': dict({
        'a': 2,
        'b': 2,
      }),
      'why': False,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 5,
      'test': dict({
        'a': 2,
        'b': 2,
      }),
      'why': True,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 5,
      'test': dict({
        'a': 2,
        'b': 2,
      }),
      'why': False,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 4,
      'test': dict({
        'a': 2,
        'b': 3,
      }),
      'why': True,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 4,
      'test': dict({
        'a': 2,
        'b': 3,
      }),
      'why': False,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 5,
      'test': dict({
        'a': 2,
        'b': 3,
      }),
      'why': True,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 5,
      'test': dict({
        'a': 2,
        'b': 3,
      }),
      'why': False,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 4,
      'test': dict({
        'a': 3,
        'b': 1,
      }),
      'why': True,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 4,
      'test': dict({
        'a': 3,
        'b': 1,
      }),
      'why': False,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 5,
      'test': dict({
        'a': 3,
        'b': 1,
      }),
      'why': True,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 5,
      'test': dict({
        'a': 3,
        'b': 1,
      }),
      'why': False,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 4,
      'test': dict({
        'a': 3,
        'b': 2,
      }),
      'why': True,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 4,
      'test': dict({
        'a': 3,
        'b': 2,
      }),
      'why': False,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 5,
      'test': dict({
        'a': 3,
        'b': 2,
      }),
      'why': True,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 5,
      'test': dict({
        'a': 3,
        'b': 2,
      }),
      'why': False,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 4,
      'test': dict({
        'a': 3,
        'b': 3,
      }),
      'why': True,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 4,
      'test': dict({
        'a': 3,
        'b': 3,
      }),
      'why': False,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 5,
      'test': dict({
        'a': 3,
        'b': 3,
      }),
      'why': True,
    }),
    dict({
      'huh': 5,
      'no': 6,
      'some': 5,
      'test': dict({
        'a': 3,
        'b': 3,
      }),
      'why': False,
    }),
  ])
# ---
